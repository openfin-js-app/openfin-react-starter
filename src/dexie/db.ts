import Dexie from 'dexie';

import { IConfigDexie } from '../reduxs';

class OpenfinReactDatabase extends Dexie{
    configs:Dexie.Table<IConfigDexie,number>;

    constructor(databaseName){
        super(databaseName);
        this.version(1).stores({
            configs:'++id,tabName,fieldName'
        })
    }
}

const db = new OpenfinReactDatabase('openfinReactDb');

db.open().catch((err)=>{
    console.error('Default openfinReactDb db failed',err);
});

export default db;