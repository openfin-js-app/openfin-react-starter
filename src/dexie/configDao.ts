import db from './db';

import { IConfigDexie } from '../redux';


export async function findAll():Promise<IConfigDexie[]>{
    return  db.table('configs').toArray();
}

export async function saveOrUpdateOneByTabNameFieldName(tabName:string, fieldName:string, value:any):Promise<IConfigDexie> {
    let one:IConfigDexie = null;
    const configs = db.table('configs');
    await db.transaction('rw',configs,async()=>{
        const founds:IConfigDexie[] = await configs
            .where({tabName,fieldName})
            .toArray();
        console.log('configDao::saveOrUpdateOneByTabNameFieldName',founds.length,founds,tabName,fieldName,value);
        if (founds.length === 0){
            one = {tabName,fieldName,value};
            const id = await configs.put(one);
            one = await configs.get(id);
        }else if (founds.length === 1){
            one = founds[0];
            await configs.update(one.id,{value});
            one = await configs.get(one.id);
        }else{
            await configs.bulkDelete(founds.map(one => one.id));
            one = {tabName,fieldName,value};
            const id = await configs.put(one);
            one = await configs.get(id);
        }
    });
    return one;
}