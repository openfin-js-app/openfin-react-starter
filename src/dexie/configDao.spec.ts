import { findAll, saveOrUpdateOneByTabNameFieldName } from './configDao';
import db from './db';

jest.mock('./db');

describe('ConfigDao',()=>{

    it('findAll async',async ()=>{
        const founds = await findAll();
        expect(founds).toMatchSnapshot();
    })


    it('saveOrUpdateOneByTabNameFieldName async',async ()=>{
        const saved = await saveOrUpdateOneByTabNameFieldName(
            'tabName', 'fieldName', 'value'
        );
        expect(saved).toMatchSnapshot();
    })

    describe('with preset',()=>{
        beforeAll(()=>{
            const configs = db.table('configs');
            configs.put({
                tabName:'tabName', fieldName:'fieldName', value:'value',
            });
            configs.put({
                tabName:'tabName', fieldName:'fieldName', value:'value',
            });
        })

        it('saveOrUpdateOneByTabNameFieldName async with preset',async ()=>{
            const saved = await saveOrUpdateOneByTabNameFieldName(
                'tabName', 'fieldName', 'value'
            );
            expect(saved).toMatchSnapshot();
        })


        it('saveOrUpdateOneByTabNameFieldName async update',async ()=>{
            const saved = await saveOrUpdateOneByTabNameFieldName(
                'tabName', 'fieldName', 'value2'
            );
            expect(saved).toMatchSnapshot();
        })

    })

});