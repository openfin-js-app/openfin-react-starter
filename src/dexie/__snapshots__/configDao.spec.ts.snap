// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigDao findAll async 1`] = `Array []`;

exports[`ConfigDao saveOrUpdateOneByTabNameFieldName async 1`] = `
Object {
  "fieldName": "fieldName",
  "id": 1,
  "tabName": "tabName",
  "value": "value",
}
`;

exports[`ConfigDao with preset saveOrUpdateOneByTabNameFieldName async update 1`] = `
Object {
  "fieldName": "fieldName",
  "id": 4,
  "tabName": "tabName",
  "value": "value2",
}
`;

exports[`ConfigDao with preset saveOrUpdateOneByTabNameFieldName async with preset 1`] = `
Object {
  "fieldName": "fieldName",
  "id": 4,
  "tabName": "tabName",
  "value": "value",
}
`;
